pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,450 @@
+		ofClient:        client,
+		interfaceStore:  interfaceStore,
+		nodeConfig:      nodeConfig,
+		podQueue:        workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "MCPodRouteControllerForPod"),
+		gwQueue:         workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "MCPodRouteControllerForGateway"),
+		gwInformer:      gwInformer.Informer(),
+		gwLister:        gwInformer.Lister(),
+		podNameIPCache:  make(map[string]string),
+		podWorkerStopCh: make(chan struct{}),
+	}
+
+	controller.gwInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueueGateway(cur)
+			},
+			// Gateway UPDATE event doesn't impact Pod flows, so ignore it.
+			DeleteFunc: func(old interface{}) {
+				controller.enqueueGateway(old)
+			},
+		},
+		resyncPeriod,
+	)
+	return controller
+}
+
+func podIPIndexFunc(obj interface{}) ([]string, error) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		return nil, fmt.Errorf("obj is not Pod: %+v", obj)
+	}
+	if pod.Status.PodIP != "" {
+		return []string{pod.Status.PodIP}, nil
+	}
+	return []string{}, nil
+}
+
+func (c *MCPodRouteController) createPodInformer() {
+	listOptions := func(options *metav1.ListOptions) {
+		options.FieldSelector = fields.OneTermNotEqualSelector("spec.nodeName", c.nodeConfig.Name).String()
+	}
+	c.podInformer = coreinformers.NewFilteredPodInformer(
+		c.k8sClient,
+		metav1.NamespaceAll,
+		0,
+		cache.Indexers{podIndexKey: podIPIndexFunc},
+		listOptions,
+	)
+	c.podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				c.enqueuePod(cur, false)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				c.enqueuePod(cur, false)
+			},
+			DeleteFunc: func(old interface{}) {
+				c.enqueuePod(old, true)
+			},
+		},
+		resyncPeriod,
+	)
+	c.podLister = corelisters.NewPodLister(c.podInformer.GetIndexer())
+}
+
+func (c *MCPodRouteController) enqueueGateway(obj interface{}) {
+	_, isGW := obj.(*mcv1alpha1.Gateway)
+	if !isGW {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(nil, "Received unexpected object", "object", obj)
+			return
+		}
+		_, ok = deletedState.Obj.(*mcv1alpha1.Gateway)
+		if !ok {
+			klog.ErrorS(nil, "DeletedFinalStateUnknown contains non-Gateway object", "object", deletedState.Obj)
+			return
+		}
+	}
+	c.gwQueue.Add(dummyKey)
+}
+
+func (c *MCPodRouteController) enqueuePod(obj interface{}, isDelete bool) {
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		klog.ErrorS(errors.New("received unexpected object"), "enqueuePod can't process event", "obj", obj)
+		return
+	}
+
+	podIP := pod.Status.PodIP
+	podName := types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name}
+	cachedPodIP, exists := c.getPodNameIPCache(podName)
+	if !exists && (podIP == "" || pod.Spec.HostNetwork) {
```

Comment:
As Quan said you should pass both old and new Pod to this func, and check both (exists && !HostNetwork && podIP != "" && nodeIP != ""), and something like:

- !isOldValid && !newValid: return
- !isOldValid: Add(newPodIP)
- !isNewValid: Add(oldPodIP)
- if oldPodIP != newPodIP: Add(newPodIP) Add(newPodIP)
- if oldNodeIP != newNodeIP: Add(newPodIP)

Reply:
Done.